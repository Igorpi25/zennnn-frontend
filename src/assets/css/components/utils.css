.overflow-scroll-touch {
  -webkit-overflow-scrolling: touch;
}

.text-overflow-ellipsis {
  text-overflow: ellipsis;
}

/* Full height dialog override */
.dialog-full-height {
  max-height: 100% !important;
}

/* Temp class */
.input-transparent {
  background: transparent!important;
}

/* <-- DOTS */
.dots {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: clip;
}
.dots::before {
  display: block;
  content:
    ". . . . . . . . . . . . . . . . . . . . "
    ". . . . . . . . . . . . . . . . . . . . "
    ". . . . . . . . . . . . . . . . . . . . "
    ". . . . . . . . . . . . . . . . . . . . "
  ;
}
/* DOTS --> */
