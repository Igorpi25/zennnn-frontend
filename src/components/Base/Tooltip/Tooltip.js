import {
  h,
  ref,
  reactive,
  computed,
  watch,
} from 'vue'

import {
  usePopperProps,
  usePopper,
  useActivatorProps,
  useActivator,
  useLazyContentProps,
  useLazyContent,
  ClickOutside,
} from 'uipart'

import uid from '../../../utils/uid'

import './Tooltip.css'

export default {
  name: 'Tooltip',

  props: {
    ...usePopperProps(),
    ...useActivatorProps(),
    ...useLazyContentProps(),
    tag: {
      type: String,
      default: 'span',
    },
    contentClass: {
      type: String,
      default: '',
    },
    transition: {
      type: [String, Object],
      default: {
        enterActiveClass: 'transition ease-out duration-150',
        enterFromClass: 'transform opacity-0 scale-50',
        enterToClass: 'transform opacity-100 scale-100',
        leaveActiveClass: 'transition ease-out duration-75',
        leaveFromClass: 'transform opacity-100',
        leaveToClass: 'transform opacity-0',
      },
    },
  },

  emits: ['update:modelValue'],

  setup (props, { slots, emit }) {
    const id = uid('tooltip-')
    const rootElement = ref(null)

    const {
      isVisible,
      isContentVisible,
      wrapperElement,
      create: createPopper,
      destroy: destroyPopper,
      genBox,
      genWrapper,
    } = usePopper(props)

    const {
      isActive,
      activatorElement,
      genActivator,
      getActivator,
    } = useActivator(props, { slots, emit })

    const lazyContentProps = reactive({
      isActive: isActive,
      disabled: computed(() => props.disabled),
    })
    const {
      showLazyContent,
    } = useLazyContent(lazyContentProps)

    const activatorAttrs = computed(() => {
      return {
        'aria-describedby': isActive.value ? id : undefined,
        'aria-controls': isActive.value ? id : undefined,
      }
    })

    const activatorListeners = computed(() => {
      return {
        onBlur: () => {
          isActive.value = false
        },
      }
    })

    watch(isActive, (val) => {
      if (props.disabled) return
      // show popper root from isActive, hide after transition end
      if (val) isVisible.value = val
      val ? activate() : deactivate()
    })

    watch(isVisible, (val) => {
      if (val !== isActive.value) {
        isActive.value = val
      }
    })

    watch(() => props.activator, () => {
      destroyPopper()
    })

    const activate = () => {
      getActivator()
      requestAnimationFrame(() => {
        isContentVisible.value = true
        createPopper(activatorElement.value, wrapperElement.value)
      })
    }

    const deactivate = () => {
      isContentVisible.value = false
    }

    const genContent = () => {
      return h('div', {
        class: {
          tooltip__content: true,
          [props.contentClass.trim()]: true,
        },
      }, slots.default ? slots.default() : undefined)
    }

    const genPopper = () => {
      const children = genBox(
        {
          id,
          role: 'tooltip',
          tabindex: '-1',
          class: 'tooltip__box',
        },
        [genContent()],
      )

      const directives = [
        [
          ClickOutside,
          {
            handler: () => {
              isActive.value = false
            },
            closeConditional: (e) => {
              const target = e.target
              return isActive.value &&
                !(wrapperElement.value && wrapperElement.value.contains(target))
            },
            include: () => [rootElement.value],
          },
        ],
      ]

      return genWrapper(
        undefined,
        children,
        directives,
      )
    }

    return {
      isActive,
      isVisible,
      isContentVisible,
      rootElement,
      activatorAttrs,
      activatorListeners,
      genPopper,
      showLazyContent,
      genActivator,
    }
  },

  render () {
    return h(this.tag, {
      ref: 'rootElement',
      class: {
        tooltip: true,
      },
    }, [
      this.showLazyContent(this.genPopper),
      this.genActivator(this.activatorAttrs, this.activatorListeners),
    ])
  },
}
