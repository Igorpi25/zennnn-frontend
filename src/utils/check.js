/**
 * String type check.
 * @param {any} v
 * @returns {boolean}
 */
export const isString = v => typeof v === 'string' || v instanceof String

/**
 * Number type check.
 * @param {any} v
 * @returns {boolean}
 */
export const isNumber = v => typeof v === 'number' && isFinite(v)

/**
 * Array type check.
 * @param {any} v
 * @returns {boolean}
 */
export const isArray = v => v && typeof v === 'object' && v.constructor === Array

/**
 * Object type check.
 * @param {any} v
 * @returns {boolean}
 */
export const isObject = v => v && typeof v === 'object' && v.constructor === Object
