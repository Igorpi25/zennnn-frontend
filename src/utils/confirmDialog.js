/**
 * Confirm
 * TODO: move to plugin
 * @returns {string} message
 */
export const confirmDialog = (msg) => {
  return new Promise((resolve, reject) => {
    const confirmed = window.confirm(msg)
    return confirmed ? resolve('confirmed') : reject('not_confirmed') // eslint-disable-line
  })
}

export default confirmDialog
