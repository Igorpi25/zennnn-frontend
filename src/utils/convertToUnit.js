/**
 * Convert to unit.
 * @param {string,number,null,undefined} value The value to convert.
 * @param {string} [unit=px] The unit value.
 * @returns {string,undefined} Returns the converted value.
 */
export const convertToUnit = (value, unit = 'px') => {
  if (value == null || value === '') {
    return undefined
  } else if (isNaN(+value)) {
    return String(value)
  } else {
    return `${Number(value)}${unit}`
  }
}

export default convertToUnit
