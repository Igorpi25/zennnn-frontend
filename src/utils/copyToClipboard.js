/**
 * Copy text to clipboard
 * @param {string} text value to copy
 */
export const copyToClipboard = (text) => {
  return new Promise((resolve, reject) => {
    const el = document.createElement('textarea')
    el.value = text
    el.setAttribute('readonly', '')
    el.style.position = 'absolute'
    el.style.left = '-9999px'
    document.body.appendChild(el)
    let selected

    try {
      // Check if there is any content selected previously
      // Store selection if found
      // Mark as false to know no selection existed before
      // eslint-disable-next-line
      document.getSelection().rangeCount > 0
        ? document.getSelection().getRangeAt(0)
        : false
      el.select()
      // Copy - only works as a result of a user action (e.g. click events)
      const successful = document.execCommand('copy')
      const msg = successful ? 'successful' : 'unsuccessful'
      resolve(msg)
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error('Unable to copy, ', error)
      reject(error)
    }

    document.body.removeChild(el)
    // If a selection existed before copying
    // Unselect everything on the HTML document
    // Restore the original selection
    if (selected) {
      document.getSelection().removeAllRanges()
      document.getSelection().addRange(selected)
    }
  })
}

export default copyToClipboard
