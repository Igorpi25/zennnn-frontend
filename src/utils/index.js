export * from './check'
export * from './convertToUnit'
export * from './deepEqual'
export * from './defaultFilter'
export * from './dom'
export * from './formatNumber'
export * from './object'
export * from './parseEventName'
export * from './uid'
export * from './upperFirst'
export * from './wait'
export * from './wrapInArray'
