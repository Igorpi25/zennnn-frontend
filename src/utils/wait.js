/**
 * Promise delay.
 * @param {number} ms delay
 * @returns {Promise}
 */
export const wait = ms => new Promise(resolve => setTimeout(resolve, ms))

export default wait
