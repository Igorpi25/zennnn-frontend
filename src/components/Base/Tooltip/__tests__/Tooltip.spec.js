import { mount } from '@vue/test-utils'
import { h, nextTick } from 'vue'

import Tooltip from '../Tooltip'

import { wait } from '../../../../utils/wait'

jest.mock('../../../../utils/uid')

describe('Tooltip.js', () => {
  let mountFunction

  beforeEach(() => {
    document.body.innerHTML = '<div id="app"></div>'

    mountFunction = (options = {}) => {
      return mount(Tooltip, {
        ...options,
      })
    }
  })

  // Render

  it('should render component and match snapshot', async () => {
    const wrapper = mountFunction({
      slots: {
        default: () => Array.from(Array(3).keys()).map(() => 'Content').join(' '),
        activator: () => h('button', 'Activator'),
      },
    })

    await wrapper.setProps({ modelValue: true })
    await wait(50)

    expect(wrapper.html()).toMatchSnapshot()
    expect(document.body.innerHTML).toMatchSnapshot()
  })

  it('should not attach content and match snapshot', async () => {
    const wrapper = mountFunction({
      props: {
        modelValue: true,
        attach: false,
      },
      slots: {
        default: () => h('span', 'Content'),
        activator: () => h('button', 'Activator'),
      },
    })

    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with min/max width and match snapshot', async () => {
    const wrapper = mountFunction({
      props: {
        modelValue: true,
        attach: false,
        minWidth: 100,
        maxWidth: 200,
      },
      slots: {
        default: () => h('span', 'Content'),
        activator: () => h('button', 'Activator'),
      },
    })

    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with zIndex prop and match snapshot', async () => {
    const wrapper = mountFunction({
      props: {
        modelValue: true,
        attach: false,
        zIndex: 10,
      },
      slots: {
        default: () => h('span', 'Content'),
        activator: () => h('button', 'Activator'),
      },
    })

    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should open on mousenter and close on mouseleave', async () => {
    const wrapper = mountFunction({
      slots: {
        default: () => h('span', 'Content'),
        activator: () => h('button', 'Activator'),
      },
    })

    const activator = wrapper.get('button')
    await activator.trigger('mouseenter')
    expect(wrapper.vm.isActive).toBe(true)
    await activator.trigger('mouseleave')
    expect(wrapper.vm.isActive).toBe(false)
  })

  it('should open on focus and close on blur', async () => {
    const wrapper = mountFunction({
      slots: {
        default: () => h('span', 'Content'),
        activator: () => h('button', 'Activator'),
      },
    })

    const activator = wrapper.get('button')
    await activator.trigger('focus')
    expect(wrapper.vm.isActive).toBe(true)
    await activator.trigger('blur')
    expect(wrapper.vm.isActive).toBe(false)
  })
})
