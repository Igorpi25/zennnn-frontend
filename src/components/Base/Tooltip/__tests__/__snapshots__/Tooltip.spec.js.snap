// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip.js should not attach content and match snapshot 1`] = `
<span class="tooltip"><!--teleport start--><div style="z-index: 10; display: none;" data-popper-root=""><transition-stub><div id="tooltip-1" role="tooltip" tabindex="-1" class="popper__box tooltip__box" style="display: none;"><div class="tooltip__content"><span>Content</span></div>
<div class="popper__arrow" data-popper-arrow=""></div>
</div>
</transition-stub>
</div>
<!--teleport end--><button role="button" aria-haspopup="true" aria-expanded="true" aria-describedby="tooltip-1" aria-controls="tooltip-1">Activator</button></span>
`;

exports[`Tooltip.js should render component and match snapshot 1`] = `<span class="tooltip"><!--teleport start--><!--teleport end--><button role="button" aria-haspopup="true" aria-expanded="true" aria-describedby="tooltip-1" aria-controls="tooltip-1">Activator</button></span>`;

exports[`Tooltip.js should render component and match snapshot 2`] = `
<div id="app">
  <div style="z-index: 10; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 2px);" data-popper-root="" data-popper-reference-hidden="" data-popper-escaped="" data-popper-placement="top">
    <transition-stub>
      <div id="tooltip-1" role="tooltip" tabindex="-1" class="popper__box tooltip__box" style="">
        <div class="tooltip__content">Content Content Content</div>
        <div class="popper__arrow" data-popper-arrow="" style="position: absolute; left: 0px; transform: translate(8px, 0px);" data-hide=""></div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`Tooltip.js should render component with min/max width and match snapshot 1`] = `
<span class="tooltip"><!--teleport start--><div style="z-index: 10; display: none;" data-popper-root=""><transition-stub><div id="tooltip-1" role="tooltip" tabindex="-1" class="popper__box tooltip__box" style="max-width: 200px; min-width: 100px; display: none;"><div class="tooltip__content"><span>Content</span></div>
<div class="popper__arrow" data-popper-arrow=""></div>
</div>
</transition-stub>
</div>
<!--teleport end--><button role="button" aria-haspopup="true" aria-expanded="true" aria-describedby="tooltip-1" aria-controls="tooltip-1">Activator</button></span>
`;

exports[`Tooltip.js should render component with zIndex prop and match snapshot 1`] = `
<span class="tooltip"><!--teleport start--><div style="z-index: 10; display: none;" data-popper-root=""><transition-stub><div id="tooltip-1" role="tooltip" tabindex="-1" class="popper__box tooltip__box" style="display: none;"><div class="tooltip__content"><span>Content</span></div>
<div class="popper__arrow" data-popper-arrow=""></div>
</div>
</transition-stub>
</div>
<!--teleport end--><button role="button" aria-haspopup="true" aria-expanded="true" aria-describedby="tooltip-1" aria-controls="tooltip-1">Activator</button></span>
`;
